File: scripts/cams-sync/sync-cams.js
#!/usr/bin/env node
/**
 * CAMS 품의서/전표 동기화 스크립트
 *
 * 사내 서버에서 실행하여 CAMS 데이터를 Railway API로 푸시합니다.
 *
 * 사용법:
 *   node sync-cams.js
 *
 * 환경변수:
 *   CAMS_BASE_URL     - CAMS 서버 URL (기본: http://cn.icams.co.kr)
 *   WORKWORK_API_URL  - WorkWork API URL (예: https://api.workwork.kr)
 *   CAMS_SYNC_API_KEY - 동기화 API 키
 *
 * 매일 아침 cron으로 실행:
 *   0 7 * * * /usr/bin/node /path/to/sync-cams.js >> /var/log/cams-sync.log 2>&1
 */

const CAMS_BASE_URL = process.env.CAMS_BASE_URL || 'http://cn.icams.co.kr';
const WORKWORK_API_URL = process.env.WORKWORK_API_URL || 'https://api.workwork.kr';
const CAMS_SYNC_API_KEY = process.env.CAMS_SYNC_API_KEY;

if (!CAMS_SYNC_API_KEY) {
  console.error('ERROR: CAMS_SYNC_API_KEY 환경변수가 필요합니다');
  process.exit(1);
}

// EUC-KR 디코딩을 위한 iconv-lite (필요시 설치: npm install iconv-lite)
let iconv;
try {
  iconv = require('iconv-lite');
} catch (e) {
  console.warn('WARN: iconv-lite 없음, UTF-8로만 디코딩합니다. EUC-KR 페이지는 깨질 수 있습니다.');
  console.warn('      npm install iconv-lite 로 설치하세요.');
}

/**
 * CAMS 페이지를 가져와서 HTML 문자열로 반환
 */
async function fetchCamsHtml(url) {
  const res = await fetch(url, {
    headers: {
      'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
      'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
      'Accept-Language': 'ko-KR,ko;q=0.9,en;q=0.8',
    },
  });

  if (!res.ok) {
    throw new Error(`HTTP ${res.status}: ${url}`);
  }

  const buf = Buffer.from(await res.arrayBuffer());

  // EUC-KR 디코딩 시도
  if (iconv) {
    const head = buf.slice(0, 1024).toString('ascii');
    const isEucKr = /charset\s*=\s*["']?(euc-kr|ks_c_5601)/i.test(head);
    if (isEucKr) {
      return iconv.decode(buf, 'euc-kr');
    }
  }

  return buf.toString('utf8');
}

/**
 * HTML에서 DataGrid span 파싱
 * <span id="gridId_lblFIELD_N">value</span> 형태
 */
function parseGrids(html) {
  const re = /<span[^>]*\bid=["']?([A-Za-z0-9]+?)_lbl([A-Za-z]+)_?(\d+)["']?[^>]*>([\s\S]*?)<\/span>/gi;
  const byGrid = {};
  let m;

  while ((m = re.exec(html)) !== null) {
    const gridId = m[1];
    const field = m[2].toLowerCase();
    const idx = Number(m[3]);
    const innerHtml = m[4];
    const text = stripTags(innerHtml).replace(/&nbsp;/gi, ' ').replace(/&amp;/gi, '&').trim();

    if (!byGrid[gridId]) byGrid[gridId] = { fieldOrder: [], rows: {} };
    const g = byGrid[gridId];
    if (!g.fieldOrder.includes(field)) g.fieldOrder.push(field);
    if (!g.rows[idx]) g.rows[idx] = {};
    g.rows[idx][field] = text;
  }

  const out = {};
  for (const [gridId, g] of Object.entries(byGrid)) {
    const indices = Object.keys(g.rows).map(Number).filter(n => Number.isFinite(n)).sort((a, b) => a - b);
    const rows = indices.map(idx => ({ _index: idx, ...g.rows[idx] }));
    out[gridId] = { id: gridId, fields: g.fieldOrder, rows };
  }
  return out;
}

function stripTags(s) {
  return String(s || '').replace(/<[^>]+>/g, '');
}

/**
 * 품의서 목록 페이지에서 품의번호 목록 추출
 */
async function fetchProposalList() {
  const url = `${CAMS_BASE_URL}/boss/mpu.aspx`;
  console.log(`[품의서] 목록 조회: ${url}`);

  const html = await fetchCamsHtml(url);
  const grids = parseGrids(html);
  const mainGrid = grids['myDataGrid'];

  if (!mainGrid || !mainGrid.rows.length) {
    console.log('[품의서] 목록이 비어있습니다');
    return [];
  }

  // slpno 필드에서 품의번호 추출
  const slpNos = mainGrid.rows
    .map(row => row.slpno)
    .filter(Boolean);

  console.log(`[품의서] ${slpNos.length}건 발견`);
  return slpNos;
}

/**
 * 품의서 상세 페이지에서 데이터 추출
 */
async function fetchProposalDetail(slpNo) {
  const url = `${CAMS_BASE_URL}/acco/mpu_list2.aspx?slp_no=${encodeURIComponent(slpNo)}`;
  console.log(`[품의서] 상세 조회: ${slpNo}`);

  const html = await fetchCamsHtml(url);
  const grids = parseGrids(html);

  // myDataGrid = 메인 정보, myDataGrid2 = 첨부파일
  const mainGrid = grids['myDataGrid'];
  const fileGrid = grids['myDataGrid2'];

  if (!mainGrid || !mainGrid.rows.length) {
    console.log(`[품의서] ${slpNo}: 상세 정보 없음`);
    return null;
  }

  const row = mainGrid.rows[0];

  // 첨부파일 처리
  const attachments = [];
  if (fileGrid && fileGrid.rows.length > 0) {
    for (let i = 0; i < fileGrid.rows.length; i++) {
      const fileRow = fileGrid.rows[i];
      const seq = i + 1;
      const filename = fileRow.title || `파일${seq}`;
      const downloadUrl = `${CAMS_BASE_URL}/acco/mpu_list2.aspx?slp_no=${encodeURIComponent(slpNo)}&sort=${seq}`;
      attachments.push({ seq, filename, downloadUrl });
    }
  }

  return {
    slpNo: row.slpno || slpNo,
    title: row.title || '',
    purpose: row.aim || '',
    drafter: row.bscname || '',
    draftDate: row.pymd || '',
    dueDate: row.wymd || '',
    amount: row.samt || '',
    paymentTerm: row.gjo || '',
    vendor: '', // 여러 필드에 흩어져 있을 수 있음
    content: row.gjo || '', // 내용이 gjo에 있는 경우
    attachments,
  };
}

/**
 * 전표 목록 페이지에서 전표번호 목록 추출
 */
async function fetchVoucherList() {
  const url = `${CAMS_BASE_URL}/boss/macco.aspx`;
  console.log(`[전표] 목록 조회: ${url}`);

  const html = await fetchCamsHtml(url);
  const grids = parseGrids(html);
  const mainGrid = grids['myDataGrid'];

  if (!mainGrid || !mainGrid.rows.length) {
    console.log('[전표] 목록이 비어있습니다');
    return [];
  }

  const slpNos = mainGrid.rows
    .map(row => row.slpno)
    .filter(Boolean);

  console.log(`[전표] ${slpNos.length}건 발견`);
  return slpNos;
}

/**
 * 전표 상세 페이지에서 데이터 추출
 */
async function fetchVoucherDetail(slpNo) {
  const url = `${CAMS_BASE_URL}/acco/macco_list.aspx?slp_no=${encodeURIComponent(slpNo)}`;
  console.log(`[전표] 상세 조회: ${slpNo}`);

  const html = await fetchCamsHtml(url);
  const grids = parseGrids(html);

  const mainGrid = grids['myDataGrid'] || grids['myDataGrid2'];

  if (!mainGrid || !mainGrid.rows.length) {
    console.log(`[전표] ${slpNo}: 상세 정보 없음`);
    return null;
  }

  const row = mainGrid.rows[0];

  return {
    slpNo: row.slpno || slpNo,
    title: row.aspnote || row.title || '',
    drafter: row.sname || '',
    draftDate: row.pymd || '',
    amount: row.amt || '',
    status: row.status || '',
    txType: row.txtype || '',
    content: '',
    attachments: [], // 전표는 첨부파일 구조가 다를 수 있음
  };
}

/**
 * WorkWork API로 데이터 푸시
 */
async function pushToWorkWork(type, items) {
  const url = `${WORKWORK_API_URL}/api/cams/sync`;
  console.log(`[푸시] ${type} ${items.length}건 → ${url}`);

  const res = await fetch(url, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'X-CAMS-API-KEY': CAMS_SYNC_API_KEY,
    },
    body: JSON.stringify({ type, items }),
  });

  if (!res.ok) {
    const text = await res.text();
    throw new Error(`푸시 실패: HTTP ${res.status} - ${text}`);
  }

  const result = await res.json();
  console.log(`[푸시] 결과:`, result);
  return result;
}

/**
 * 메인 동기화 실행
 */
async function main() {
  console.log('========================================');
  console.log(`CAMS 동기화 시작: ${new Date().toISOString()}`);
  console.log(`CAMS: ${CAMS_BASE_URL}`);
  console.log(`API: ${WORKWORK_API_URL}`);
  console.log('========================================');

  try {
    // 품의서 동기화
    const proposalSlpNos = await fetchProposalList();
    const proposals = [];

    for (const slpNo of proposalSlpNos) {
      try {
        const detail = await fetchProposalDetail(slpNo);
        if (detail) proposals.push(detail);
        // 요청 간 딜레이 (CAMS 서버 부하 방지)
        await new Promise(r => setTimeout(r, 500));
      } catch (e) {
        console.error(`[품의서] ${slpNo} 조회 실패:`, e.message);
      }
    }

    if (proposals.length > 0) {
      await pushToWorkWork('proposals', proposals);
    }

    // 전표 동기화
    const voucherSlpNos = await fetchVoucherList();
    const vouchers = [];

    for (const slpNo of voucherSlpNos) {
      try {
        const detail = await fetchVoucherDetail(slpNo);
        if (detail) vouchers.push(detail);
        await new Promise(r => setTimeout(r, 500));
      } catch (e) {
        console.error(`[전표] ${slpNo} 조회 실패:`, e.message);
      }
    }

    if (vouchers.length > 0) {
      await pushToWorkWork('vouchers', vouchers);
    }

    console.log('========================================');
    console.log(`동기화 완료: 품의서 ${proposals.length}건, 전표 ${vouchers.length}건`);
    console.log('========================================');

  } catch (e) {
    console.error('동기화 실패:', e);
    process.exit(1);
  }
}

main();

File: scripts/cams-sync/sync_cams.py
#!/usr/bin/env python3
"""
CAMS 품의서/전표 동기화 스크립트 (Python)

사내 PC에서 실행하여 CAMS 데이터를 WorkWork API로 푸시합니다.

사용법:
    python sync_cams.py

환경변수 또는 아래 설정값 직접 수정:
    CAMS_BASE_URL     - CAMS 서버 URL
    WORKWORK_API_URL  - WorkWork API URL
    CAMS_SYNC_API_KEY - 동기화 API 키
"""

import os
import re
import time
import requests
from datetime import datetime

# ============================================================
# 설정 (환경변수 또는 직접 입력)
# ============================================================
CAMS_BASE_URL = os.getenv('CAMS_BASE_URL', 'http://cn.icams.co.kr')
WORKWORK_API_URL = os.getenv('WORKWORK_API_URL', 'https://api.workwork.kr')
CAMS_SYNC_API_KEY = os.getenv('CAMS_SYNC_API_KEY', '여기에_API_키_입력')

# ============================================================
# HTML 파싱 함수
# ============================================================

def fetch_cams_html(url: str) -> str:
    """CAMS 페이지를 가져와서 HTML 문자열로 반환"""
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        'Accept-Language': 'ko-KR,ko;q=0.9,en;q=0.8',
    }

    res = requests.get(url, headers=headers, timeout=30)
    res.raise_for_status()

    # EUC-KR 디코딩 시도
    content_type = res.headers.get('content-type', '')
    if 'euc-kr' in content_type.lower() or 'ks_c_5601' in content_type.lower():
        res.encoding = 'euc-kr'
    else:
        # HTML 내 charset 확인
        if b'euc-kr' in res.content[:1024].lower() or b'ks_c_5601' in res.content[:1024].lower():
            res.encoding = 'euc-kr'

    return res.text


def strip_tags(html: str) -> str:
    """HTML 태그 제거"""
    return re.sub(r'<[^>]+>', '', html or '')


def decode_entities(text: str) -> str:
    """HTML 엔티티 디코딩"""
    text = text.replace('&nbsp;', ' ')
    text = text.replace('&amp;', '&')
    text = text.replace('&lt;', '<')
    text = text.replace('&gt;', '>')
    text = text.replace('&quot;', '"')
    text = text.replace('&#39;', "'")
    return text


def parse_grids(html: str) -> dict:
    """
    HTML에서 DataGrid span 파싱
    <span id="gridId_lblFIELD_N">value</span> 형태
    """
    pattern = r'<span[^>]*\bid=["\']?([A-Za-z0-9]+?)_lbl([A-Za-z]+)_?(\d+)["\']?[^>]*>([\s\S]*?)</span>'

    by_grid = {}

    for match in re.finditer(pattern, html, re.IGNORECASE):
        grid_id = match.group(1)
        field = match.group(2).lower()
        idx = int(match.group(3))
        inner_html = match.group(4)
        text = decode_entities(strip_tags(inner_html)).strip()

        if grid_id not in by_grid:
            by_grid[grid_id] = {'field_order': [], 'rows': {}}

        g = by_grid[grid_id]
        if field not in g['field_order']:
            g['field_order'].append(field)
        if idx not in g['rows']:
            g['rows'][idx] = {}
        g['rows'][idx][field] = text

    result = {}
    for grid_id, g in by_grid.items():
        indices = sorted([i for i in g['rows'].keys() if isinstance(i, int)])
        rows = [{'_index': idx, **g['rows'][idx]} for idx in indices]
        result[grid_id] = {
            'id': grid_id,
            'fields': g['field_order'],
            'rows': rows
        }

    return result


# ============================================================
# 품의서 처리
# ============================================================

def fetch_proposal_list() -> list:
    """품의서 목록 페이지에서 품의번호 목록 추출"""
    url = f"{CAMS_BASE_URL}/boss/mpu.aspx"
    print(f"[품의서] 목록 조회: {url}")

    html = fetch_cams_html(url)
    grids = parse_grids(html)
    main_grid = grids.get('myDataGrid')

    if not main_grid or not main_grid['rows']:
        print("[품의서] 목록이 비어있습니다")
        return []

    slp_nos = [row.get('slpno') for row in main_grid['rows'] if row.get('slpno')]
    print(f"[품의서] {len(slp_nos)}건 발견")
    return slp_nos


def fetch_proposal_detail(slp_no: str) -> dict:
    """품의서 상세 페이지에서 데이터 추출"""
    url = f"{CAMS_BASE_URL}/acco/mpu_list2.aspx?slp_no={slp_no}"
    print(f"[품의서] 상세 조회: {slp_no}")

    html = fetch_cams_html(url)
    grids = parse_grids(html)

    main_grid = grids.get('myDataGrid')
    file_grid = grids.get('myDataGrid2')

    if not main_grid or not main_grid['rows']:
        print(f"[품의서] {slp_no}: 상세 정보 없음")
        return None

    row = main_grid['rows'][0]

    # 첨부파일 처리
    attachments = []
    if file_grid and file_grid['rows']:
        for i, file_row in enumerate(file_grid['rows']):
            seq = i + 1
            filename = file_row.get('title', f'파일{seq}')
            download_url = f"{CAMS_BASE_URL}/acco/mpu_list2.aspx?slp_no={slp_no}&sort={seq}"
            attachments.append({
                'seq': seq,
                'filename': filename,
                'downloadUrl': download_url
            })

    return {
        'slpNo': row.get('slpno', slp_no),
        'title': row.get('title', ''),
        'purpose': row.get('aim', ''),
        'drafter': row.get('bscname', ''),
        'draftDate': row.get('pymd', ''),
        'dueDate': row.get('wymd', ''),
        'amount': row.get('samt', ''),
        'paymentTerm': row.get('gjo', ''),
        'vendor': '',
        'content': row.get('gjo', ''),
        'attachments': attachments
    }


# ============================================================
# 전표 처리
# ============================================================

def fetch_voucher_list() -> list:
    """전표 목록 페이지에서 전표번호 목록 추출"""
    url = f"{CAMS_BASE_URL}/boss/macco.aspx"
    print(f"[전표] 목록 조회: {url}")

    html = fetch_cams_html(url)
    grids = parse_grids(html)
    main_grid = grids.get('myDataGrid')

    if not main_grid or not main_grid['rows']:
        print("[전표] 목록이 비어있습니다")
        return []

    slp_nos = [row.get('slpno') for row in main_grid['rows'] if row.get('slpno')]
    print(f"[전표] {len(slp_nos)}건 발견")
    return slp_nos


def fetch_voucher_detail(slp_no: str) -> dict:
    """전표 상세 페이지에서 데이터 추출"""
    url = f"{CAMS_BASE_URL}/acco/macco_list.aspx?slp_no={slp_no}"
    print(f"[전표] 상세 조회: {slp_no}")

    html = fetch_cams_html(url)
    grids = parse_grids(html)

    main_grid = grids.get('myDataGrid') or grids.get('myDataGrid2')

    if not main_grid or not main_grid['rows']:
        print(f"[전표] {slp_no}: 상세 정보 없음")
        return None

    row = main_grid['rows'][0]

    return {
        'slpNo': row.get('slpno', slp_no),
        'title': row.get('aspnote', '') or row.get('title', ''),
        'drafter': row.get('sname', ''),
        'draftDate': row.get('pymd', ''),
        'amount': row.get('amt', ''),
        'status': row.get('status', ''),
        'txType': row.get('txtype', ''),
        'content': '',
        'attachments': []
    }


# ============================================================
# API 푸시
# ============================================================

def push_to_workwork(data_type: str, items: list) -> dict:
    """WorkWork API로 데이터 푸시"""
    url = f"{WORKWORK_API_URL}/api/cams/sync"
    print(f"[푸시] {data_type} {len(items)}건 → {url}")

    headers = {
        'Content-Type': 'application/json',
        'X-CAMS-API-KEY': CAMS_SYNC_API_KEY
    }

    payload = {
        'type': data_type,
        'items': items
    }

    res = requests.post(url, json=payload, headers=headers, timeout=60)

    if not res.ok:
        raise Exception(f"푸시 실패: HTTP {res.status_code} - {res.text}")

    result = res.json()
    print(f"[푸시] 결과: {result}")
    return result


# ============================================================
# 메인
# ============================================================

def main():
    print("=" * 50)
    print(f"CAMS 동기화 시작: {datetime.now().isoformat()}")
    print(f"CAMS: {CAMS_BASE_URL}")
    print(f"API: {WORKWORK_API_URL}")
    print("=" * 50)

    if CAMS_SYNC_API_KEY == '여기에_API_키_입력':
        print("ERROR: CAMS_SYNC_API_KEY를 설정해주세요!")
        return

    try:
        # 품의서 동기화
        proposal_slp_nos = fetch_proposal_list()
        proposals = []

        for slp_no in proposal_slp_nos:
            try:
                detail = fetch_proposal_detail(slp_no)
                if detail:
                    proposals.append(detail)
                time.sleep(0.5)  # 요청 간 딜레이
            except Exception as e:
                print(f"[품의서] {slp_no} 조회 실패: {e}")

        if proposals:
            push_to_workwork('proposals', proposals)

        # 전표 동기화
        voucher_slp_nos = fetch_voucher_list()
        vouchers = []

        for slp_no in voucher_slp_nos:
            try:
                detail = fetch_voucher_detail(slp_no)
                if detail:
                    vouchers.append(detail)
                time.sleep(0.5)
            except Exception as e:
                print(f"[전표] {slp_no} 조회 실패: {e}")

        if vouchers:
            push_to_workwork('vouchers', vouchers)

        print("=" * 50)
        print(f"동기화 완료: 품의서 {len(proposals)}건, 전표 {len(vouchers)}건")
        print("=" * 50)

    except Exception as e:
        print(f"동기화 실패: {e}")
        raise


if __name__ == '__main__':
    main()
